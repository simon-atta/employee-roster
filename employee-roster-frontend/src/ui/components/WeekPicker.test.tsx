/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { shallow } from 'enzyme';
import DatePicker from 'react-datepicker';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import moment from 'moment-timezone';
import 'moment/locale/en-ca';

import MockDate from 'mockdate';
import { Button, Text } from '@patternfly/react-core';
import { HistoryIcon, CalendarIcon } from '@patternfly/react-icons';
import WeekPicker, { DateRangeInputElement } from './WeekPicker';

describe('WeekPicker component', () => {
  beforeAll(() => {
    moment.locale('en');
  });

  it('should render correctly', () => {
    const weekPicker = shallow(<WeekPicker value={moment('2019-07-03').toDate()} onChange={jest.fn()} />);
    weekPicker.setState({ isOpen: true });
    expect(toJson(weekPicker)).toMatchSnapshot();
  });

  it('should pass previous week to onChange when previous week is clicked', () => {
    const onChange = jest.fn();
    const weekPicker = shallow(<WeekPicker value={moment('2019-07-03').toDate()} onChange={onChange} />);
    weekPicker.find('[aria-label="Previous Week"]').simulate('click');
    expect(onChange).toBeCalled();
    expect(onChange)
      .toBeCalledWith(moment('2019-06-23').toDate(), moment('2019-06-23').endOf('week').toDate());
  });

  it('should pass next week to onChange when next week is clicked', () => {
    const onChange = jest.fn();
    const weekPicker = shallow(<WeekPicker value={moment('2019-07-03').toDate()} onChange={onChange} />);
    weekPicker.find('[aria-label="Next Week"]').simulate('click');
    expect(onChange).toBeCalled();
    expect(onChange)
      .toBeCalledWith(moment('2019-07-07').toDate(), moment('2019-07-07').endOf('week').toDate());
  });

  it('should go to the week containing a day when the day is clicked', () => {
    const onChange = jest.fn();
    const weekPicker = shallow(<WeekPicker value={moment('2019-07-03').toDate()} onChange={onChange} />);
    weekPicker.find(DatePicker).simulate('change', moment('2019-07-30').toDate());
    expect(onChange).toBeCalled();
    expect(onChange)
      .toBeCalledWith(moment('2019-07-28').toDate(), moment('2019-07-28').endOf('week').toDate());
  });
});

describe('WeekPicker custom input component', () => {
  beforeAll(() => {
    moment.locale('en');
  });

  it('custom input should render correctly', () => {
    const solvingStartTime = moment('2019-07-23').toDate();
    MockDate.set(solvingStartTime);
    const goToCurrentWeek = jest.fn();
    const date = moment('2019-07-03').toDate();
    const datePickerRef = { current: { setOpen: jest.fn() } };
    const customInput = shallow(
      <DateRangeInputElement
        value={date}
        datePickerRef={datePickerRef as any as React.RefObject<DatePicker>}
        goToCurrentWeek={goToCurrentWeek}
      />,
    );

    expect(customInput).toMatchSnapshot();
  });

  it('should go to current week when the reset button is clicked', () => {
    const solvingStartTime = moment('2019-07-23').toDate();
    MockDate.set(solvingStartTime);
    const goToCurrentWeek = jest.fn();
    const date = moment('2019-07-03').toDate();
    const datePickerRef = { current: { setOpen: jest.fn() } };
    const customInput = shallow(
      <DateRangeInputElement
        value={date}
        datePickerRef={datePickerRef as any as React.RefObject<DatePicker>}
        goToCurrentWeek={goToCurrentWeek}
      />,
    );

    customInput.find(Button).filterWhere(p => p.contains(<HistoryIcon />)).simulate('click');
    expect(datePickerRef.current.setOpen).not.toBeCalled();
    expect(goToCurrentWeek).toBeCalled();
  });

  it('should open the week selector if the input or calendar icon is clicked', () => {
    const solvingStartTime = moment('2019-07-23').toDate();
    MockDate.set(solvingStartTime);
    const goToCurrentWeek = jest.fn();
    const date = moment('2019-07-03').toDate();
    const datePickerRef = { current: { setOpen: jest.fn() } };
    const customInput = shallow(
      <DateRangeInputElement
        value={date}
        datePickerRef={datePickerRef as any as React.RefObject<DatePicker>}
        goToCurrentWeek={goToCurrentWeek}
      />,
    );

    customInput.find(Button).filterWhere(p => p.contains(<CalendarIcon />)).simulate('click');
    expect(datePickerRef.current.setOpen).toBeCalled();
    expect(goToCurrentWeek).not.toBeCalled();

    datePickerRef.current.setOpen.mockClear();

    customInput.find(Text).simulate('click');
    expect(datePickerRef.current.setOpen).toBeCalled();
    expect(goToCurrentWeek).not.toBeCalled();
  });
});
