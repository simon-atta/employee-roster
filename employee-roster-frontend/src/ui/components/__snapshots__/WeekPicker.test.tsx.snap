// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeekPicker component should render correctly 1`] = `
<div
  className="week-picker-container"
>
  <Button
    aria-label="Previous Week"
    onClick={[Function]}
    variant="plain"
  >
    <svg
      aria-hidden="true"
      fill="currentColor"
      height="1em"
      role="img"
      style={
        Object {
          "verticalAlign": "-0.125em",
        }
      }
      viewBox="0 0 256 512"
      width="1em"
    >
      <path
        d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
        transform=""
      />
    </svg>
  </Button>
  <t
    allowSameDay={false}
    customInput={
      <ForwardRef
        datePickerRef={
          Object {
            "current": null,
          }
        }
        goToCurrentWeek={[Function]}
      />
    }
    dateFormat="MM/dd/yyyy"
    dateFormatCalendar="LLLL yyyy"
    disabled={false}
    disabledKeyboardNavigation={false}
    dropdownMode="scroll"
    endDate={2019-07-06T23:59:59.999Z}
    inlineFocusSelectedMonth={false}
    monthsShown={1}
    nextMonthButtonLabel="Next Month"
    nextYearButtonLabel="Next Year"
    onBlur={[Function]}
    onCalendarClose={[Function]}
    onCalendarOpen={[Function]}
    onChange={[Function]}
    onClickOutside={[Function]}
    onFocus={[Function]}
    onInputClick={[Function]}
    onInputError={[Function]}
    onKeyDown={[Function]}
    onMonthChange={[Function]}
    onSelect={[Function]}
    onYearChange={[Function]}
    preventOpenOnFocus={false}
    previousMonthButtonLabel="Previous Month"
    previousYearButtonLabel="Previous Year"
    readOnly={false}
    renderDayContents={[Function]}
    required={true}
    selected={2019-06-30T00:00:00.000Z}
    shouldCloseOnSelect={true}
    showMonthYearPicker={false}
    showPopperArrow={true}
    showPreviousMonths={false}
    showQuarterYearPicker={false}
    showTimeInput={false}
    showTimeSelect={false}
    startDate={2019-06-30T00:00:00.000Z}
    strictParsing={false}
    timeCaption="Time"
    timeInputLabel="Time"
    timeIntervals={30}
    withPortal={false}
  />
  <Button
    aria-label="Next Week"
    onClick={[Function]}
    variant="plain"
  >
    <svg
      aria-hidden="true"
      fill="currentColor"
      height="1em"
      role="img"
      style={
        Object {
          "verticalAlign": "-0.125em",
        }
      }
      viewBox="0 0 256 512"
      width="1em"
    >
      <path
        d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
        transform=""
      />
    </svg>
  </Button>
</div>
`;

exports[`WeekPicker custom input component custom input should render correctly 1`] = `
<div
  className="pf-c-form-control"
  style={
    Object {
      "alignItems": "center",
      "display": "grid",
      "gridTemplateColumns": "1fr auto auto",
      "height": "auto",
    }
  }
>
  <Text
    onClick={[Function]}
  >
    06/30/2019 - 07/06/2019
  </Text>
  <Button
    onClick={[Function]}
    style={
      Object {
        "color": "black",
      }
    }
    variant="link"
  >
    <HistoryIcon
      color="currentColor"
      noVerticalAlign={false}
      size="sm"
    />
  </Button>
  <Button
    onClick={[Function]}
    style={
      Object {
        "color": "black",
      }
    }
    variant="link"
  >
    <CalendarIcon
      color="currentColor"
      noVerticalAlign={false}
      size="sm"
    />
  </Button>
</div>
`;
